adding exceptions for QueueMe

vsoch opened the pull request:
This will add an exceptions module to replace many of the sys.exits in the library. We maintain sys.exit when we are interacting with the client, however, as exit is the correct action to take.

 - This will close #34 

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch committed:
adding exceptions for QueueMe

Signed-off-by: vsoch <vsochat@stanford.edu>

yarikoptic commented on qme/exceptions.py:
```diff
@@ -0,0 +1,177 @@
+        return "Cannot find task {}.".format(self.taskid)
+
+
+class NoTasksError(RuntimeError):
```
Ideally all the Task errors should be a subclass for a base TaskError. Again, name of the exception is informative on its own

vsoch replied:
okay, will make this change (done locally).

yarikoptic commented on qme/exceptions.py:
```diff
@@ -0,0 +1,177 @@
+
+
+class UnrecognizedTargetError(RuntimeError):
+    """Thrown if an unrecognized target is given for an action
```
If given by the user than l then may be subclass of ValueError?

vsoch replied:
Fixed locally!

yarikoptic commented on qme/exceptions.py:
```diff
@@ -0,0 +1,177 @@
+
+    def __str__(self):
+        return (
+            "Database must start with sqlite, filesystem, mysql+pymysql, or postgres."
```
Hardcoded values, likely to get incorrect. Isn't there some registry to take known called from?

vsoch replied:
Sqlalchemy has a defined set https://docs.sqlalchemy.org/en/13/core/engines.html#database-urls but I've only tested (and thus would want to support) a subset of that - so this is the small set that qme supports. Would it really add value to have this list be a variable somewhere else? I only really use it here - all other checks are for just single entities (e.g., if filesystem, then do this, if postgres or sqlite, do this other thing.

yarikoptic replied:
If it could be anything what it supports, say it, may be with this set in e.g

yarikoptic replied:
By say it, I meant say so ;-)

yarikoptic replied:
But then why just not to let sqlalchemy exception to bubble up?

vsoch replied:
Sorry I don't understand, so how would you like to see this changed?

vsoch replied:
> But then why just not to let sqlalchemy exception to bubble up?

Because if they put in a prefix supported by sqlalechemy (but not my library) it wouldn't.

yarikoptic replied:
Note, docstring seems to describe something else,... will check how you use it when get back to the laptop - tricky on the phone

vsoch replied:
That's a bug! Fixed.

yarikoptic commented on qme/exceptions.py:
```diff
@@ -0,0 +1,177 @@
+
+
+class InvalidDatabaseFilename(RuntimeError):
+    """Database filename is invalid (must end in .db
```
```suggestion
    """Database filename is invalid (must end in .db)
```

yarikoptic commented on qme/exceptions.py:
```diff
@@ -0,0 +1,177 @@
+    def __str__(self):
+        return "Invalid database file for sqlite, '{}' does not end in .db".format(
+            self.filename
+        )
```
Overriding str of exceptions is a tricky business, ideally should be done with awareness of possible other arguments passed into its args, kwargs if those are allowed. Unfortunately I don't have an ultimate answer/suggestion here. I typically just complement default str of subclass with pertinent to this class information, like a filename in this case. Note that they exception name is also informative, and here message is too specific

vsoch replied:
I could just leave it the way it was - have a system exit instead. Would that be better?

yarikoptic replied:
No, SystemExit is a specific exception, not to be used unless a better exception to be raised. You could just subclass 

```
class InvalidDatabaseFilename(ValueError):
   pass
```

and be done. then when raising provide that invalid filename - `raise InvalidDatabaseFilename(filename)` or add more detail to the message. This Exception would become usable by programs which want to react, and informative to the user if it bubbles all the way up

vsoch replied:
Oh, that does work! Awesome! Fixed.

yarikoptic commented on qme/main/database/relational.py:
```diff
@@ -38,9 +43,7 @@ def __init__(self, config_dir, config=None, **kwargs):
-            sys.exit(
-                "A database url must be defined to use a relational database. Set with qme config --database"
-            )
+            raise MissingDatabaseString
```
learn something new each day.  I was certain that "legit" way is to raise instances. But it seems https://stackoverflow.com/questions/19768515/is-there-a-difference-between-raising-exception-class-and-exception-instance/19768732 that it is ok to raise class -- it will be instantiated.  I typically though just instantiate explicitly, and since typically it is a more general exception - that message is provided as an argument to the exception.

vsoch replied:
Hooray! I've seen it more commonly raised as a class, so I didn't think about this much. I think the instance would make sense if you needed to provide some argument to parse.

vsoch committed:
Merge branch 'master' into add/exceptions

vsoch committed:
fixing new unused imports for exceptions pr

Signed-off-by: vsoch <vsochat@stanford.edu>

yarikoptic commented on qme/exceptions.py:
```diff
@@ -0,0 +1,177 @@
+        return "{} {}.".format(self.name, self.reason)
+
+
+class YerAnIdiotHarryError(RuntimeError):
```
I would not have so much fun!  Exception names are information and they are API.  Should be informative and relatively "sparse", i.e. not a dedicated exception for every possible place where things could be raising.

vsoch replied:
hehe, yeah this was more of just putting in there to give you a laugh, I'll remove it :)

vsoch replied:
but seriously, if only we had these in real life! It would be much easier to raise a YouForgotToLocktheDoorAgain exception and have a large bird fly and poopland on the perpetrator than having to either lock it myself or remind said perpetrator that locking doors is a good idea. :)

yarikoptic commented on qme/exceptions.py:
```diff
@@ -0,0 +1,177 @@
+        )
+
+
+class OutputParsingError(RuntimeError):
```
also - more of `ValueError` or some other base exception?

vsoch replied:
Agree.

yarikoptic reviewed (changes requested):
I love exceptions, but care should be taken when defining them since they become part of the API etc.  It seems that currently there is way too many exceptions defined and they are all subclass of the non-specific RuntimeError, whenever I feel somce could be IOError or ValueError etc, depending on any particular case.  Also some base exception would be useful where groupping is needed (TaskError, JobError)

vsoch committed:
fixing @yarikoptic suggestions for exceptions

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch committed:
InvalidDatabaseFilename should be ValueError, and less specific

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch committed:
wrong docstring for DatabaseStringFormat error

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
@yarikoptic I think I addressed all your concerns so I'm going to merge! Open an issue when you get some time to review again if there are remaining issues.
